File: money.erl
-module(money).
-export([start/0,masterThread/3,requestFunction/3,failFunction/2,successFunction/2,approvalBank/3,denialBank/3,finalBankAmount/2]).

start() ->
  {_,Customer}= file:consult("customers.txt"),
  io:format("** Customers and loan objectives **~n"),
  startCustomerBank(Customer),
  Counter = customerCounter(Customer,0),

  {_,Bank}= file:consult("banks.txt"),
  io:format("~n~n** Banks and financial resources **~n"),
  startCustomerBank(Bank),

  register(list_to_atom("master"),spawn(money,masterThread,[0,Counter,Bank])),
  lists:map(fun(ListBank) -> master ! {ListBank} end, Bank),
  lists:map(fun(ListCustomer) -> master ! {ListCustomer,Bank,Counter} end, Customer).


masterThread(Value,CustomerCounter,BankDataList) ->
  receive
    {ElementCustomer,Bank,Counter} ->
      {Head1,Tail1} = ElementCustomer,
      register(Head1,spawn(customer,customer,[Head1,Tail1,Bank,Counter])),
      Head1 ! {Head1,Tail1,self()},
      masterThread(Value,CustomerCounter,BankDataList);
    {ElementBank} ->
      {Head1,Tail1} = ElementBank,
      register(Head1,spawn(bank,bank,[Head1,Tail1])),
      masterThread(Value,CustomerCounter,BankDataList);
    {Message, Value1, Bank_Temp_List,Temp} ->
      Value2 = Value + Value1,
      TempData = CustomerCounter,
      if
        (Value2 == TempData)->
          printBankList(self(),BankDataList);
        true ->
          io:fwrite("")
      end,
      masterThread(Value2,CustomerCounter,BankDataList)
  end.


approvalBank(Bank_name,Loan_Amount,Customer_Name)->
  io:fwrite("~p approves a loan ~p dollars from ~p~n",[Bank_name,Loan_Amount,Customer_Name]).

denialBank(Bank_name,Loan_Amount,Customer_Name)->
  io:fwrite("~p denies a loan ~p dollars from ~p~n",[Bank_name,Loan_Amount,Customer_Name]).

successFunction(CustomerName,Data)->
  io:fwrite("     ~n   ~p has reached the objective of ~p dollar(s). Woo Hoo!~n~n",[CustomerName,Data]).

failFunction(CustomerName,Data)->
  io:fwrite("     ~n   ~p was only able to borrow ~p dollar(s).Boo Hoo!~n~n",[CustomerName,Data]).

requestFunction(CustomerName, RequiredAmount, BankName) ->
  io:fwrite("~p requests a loan of ~p dollar(s) from ~p ~n",[CustomerName,RequiredAmount,BankName]).

finalBankAmount(Bank_name,Amount)->
  io:fwrite("  ~p has ~p dollar(s) remaining~n",[Bank_name,Amount]).

printBankList(Sender,[]) ->
  timer:sleep(1);

printBankList(Sender, [Head|Tail]) ->
  {Head_Temp,Tail_Temp} = Head,
  Head_Temp ! {"print",Sender,1},
  unregister(Head_Temp),
  printBankList(Sender,Tail).


customerCounter([],Counter)->
  Counter;

customerCounter([Head|Tail],Counter)->
  Temp_Counter = Counter + 1,
  customerCounter(Tail,Temp_Counter).

startCustomerBank([]) ->
  timer:sleep(1);

startCustomerBank([Head|Tail]) ->
  {Head1,Tail1} = Head,
  io:format(" ~p: ~p~n",[Head1,Tail1]),
  timer:sleep(500),
  startCustomerBank(Tail).

File: money.java
import java.io.*;
import java.util.*;

public class money {

    static ArrayList<bank> banks = new ArrayList<bank>();
    static ArrayList<customer> cus = new ArrayList<customer>();
    static ArrayList<bank> mybankObjects = new ArrayList<bank>();
    public static ArrayList<bank> getBankList()
    {
        return banks;
    }
    public static void ListOfBank(bank bankObject)
    {
        int check = 0;
        for(int i=0;i<mybankObjects.size();i++)
        {
            if(bankObject.getName().equals(mybankObjects.get(i).getName()))
            {
                check = 1;
            }
        }
        if(check == 0) {
            mybankObjects.add(bankObject);
        }
    }
    public void startReadingFile()
    {
        try {
            File file = new File("banks.txt");
            Scanner scanner = new Scanner(file);
            System.out.println("** Banks and financial resources **");
            while (scanner.hasNext()) {
                String data = scanner.next();
                String array[] = data.substring(1,data.length()-2).split("[,]");
                bank newBank = new bank(array[0],Integer.parseInt(array[1]));
                this.banks.add(newBank);
                Thread newThread = new Thread(newBank);
                newThread.start();
                newThread.sleep(1000);

            }
            scanner.close();

            file = new File("customers.txt");
            scanner = new Scanner(file);
            System.out.println("\n\n** Customers and loan objectives **");
            while (scanner.hasNext()) {
                String data = scanner.next();
                String array[] = data.substring(1,data.length()-2).split("[,]");
                customer newCustomer = new customer(array[0],Integer.parseInt(array[1]));
                this.cus.add(newCustomer);
                Thread.sleep(1000);
            }
            System.out.println("\n\n");
            for (customer j : this.cus) {
                Thread newThread = new Thread(j);
                newThread.start();

            }
            scanner.close();

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    public static void rejectLoan(String bankValue, int amount, String customerName)
    {
        System.out.println(bankValue + " denies a loan of " + amount + " from " + customerName);
    }
    public static void approvedLoan(String bankValue, int amount, String customerName)
    {
        System.out.println(bankValue + " approves a loan of " + amount + " dollars from " + customerName);
    }
    public static void reachedObjective(String customerName, int requiredAmount){
        System.out.println(customerName + " has reached the objective of " + requiredAmount + " dollar(s). Woo Hoo!");
    }
    public static void failedObjective(String customerName, int requiredAmount){
        System.out.println(customerName + " was only able to borrow " + requiredAmount + " dollar(s) Boo Hoo!");
    }
    public static void requestLoan(String customerName, int loanAmount, String Name)
    {
        System.out.println(customerName + " requests a loan of " + loanAmount + " dollars from " + Name);
    }
    public void startRunning()
    {
        while(true)
        {
            int counter = 0;
            for(int i=0;i<this.cus.size();i++)
            {
                if(this.cus.get(i).getFlag() == 1)
                {
                    counter++;
                }
            }
            if(counter == this.cus.size())
            {
                for (bank temp : this.banks) {
                    System.out.println(temp.getName() + " has " + temp.getTotalAmount() + " dollar(s) remaining.");
                }
                for (bank temp : this.mybankObjects) {
                    System.out.println(temp.getName() + " has " + temp.getTotalAmount() + " dollar(s) remaining.");
                }
                break;
            }
        }
    }
    public static void main(String args[])
    {
        money mo = new money();
        mo.startReadingFile();
        mo.startRunning();

    }
}
